### PR TITLE
Add a modified simplify function for building Hamiltonians

### PR DESCRIPTION
**Pull request by soranjh**
**Context:**
Add a modified version of the simplify function to the `hf` module. This functions combines redundant terms in a Hamiltonian and eliminates terms with zero coefficients.

**Description of the Change:**
Function `_simplify` is added to `qml.hf.hamiltonian`.

**Benefits:**
The new function makes construction of molecular Hamiltonians more efficient. For LiH, as an example, the time to construct the Hamiltonian is reduced from ~200 s to ~10 s.

**Possible Drawbacks:**

**Related GitHub Issues:**

### CODE REVIEW

File: pennylane/hf/hamiltonian.py [hunk 1]
```diff
@@ -247,7 +247,8 @@ def hamiltonian(*args):
         for n, t in enumerate(h_ferm[1]):
 
             if len(t) == 0:
-                coeffs = np.array([h_ferm[0][n]])
+                coeffs = anp.array([0.0])
+                coeffs = coeffs + np.array([h_ferm[0][n]])
```
Comment (soranjh) [inline pennylane/hf/hamiltonian.py:251]: This fixes a numpy "casting complex values to real" warning.

File: doc/releases/changelog-dev.md [hunk 2]
```diff
@@ -4,6 +4,13 @@
 
 <h3>New features since last release</h3>
 
+* Added a modified version of the `_simplify` function to the `hf` module.
+  [(#2103)](https://github.com/PennyLaneAI/pennylane/pull/2103)
+
+  This functions combines redundant terms in a Hamiltonian and eliminates terms with a zero
+  coefficient. The new function makes construction of molecular Hamiltonians more efficient.
+  For LiH, as an example, the time to construct the Hamiltonian is reduced from ~200 s to ~10 s.
```
Comment (mariaschuld) [inline doc/releases/changelog-dev.md:12]: :heart:

File: pennylane/hf/hamiltonian.py [hunk 3]
```diff
@@ -352,6 +353,54 @@ def _generate_qubit_operator(op):
...
+        ops.append(qml.grouping.string_to_pauli_word(o[i], wire_map=wiremap))
+
+    try:
+        coeffs = qml.math.stack(coeffs)
+    except ValueError:
+        pass
+
+    return qml.Hamiltonian(coeffs, ops)
+
+
```
Comment (mariaschuld) [inline pennylane/hf/hamiltonian.py:403]: It would be great to reproduce this pattern in the `Hamiltonian` class. At the moment, `H.simplify()` is a method that manipulates the internal data of `H`, which is dangerous. We'd like it to return a new operator. Another option is `qml.simplify(H)`, which could be cool because other ops may be simplified in some way too.

I don't think this relates to this PR though, which is just about building Hamiltonians in a special way, right?

File: doc/releases/changelog-dev.md (hunk 2 above)
Comment (ixfoduap) [inline doc/releases/changelog-dev.md:12]:
```suggestion
  For LiH, as an example, the time to construct the Hamiltonian is reduced roughly by a factor of 20.
```
Since runtimes depend on the device running the code

File: pennylane/hf/hamiltonian.py [hunk 4]
```diff
@@ -352,6 +353,54 @@ def _generate_qubit_operator(op):
     return c, o
 
 
+def _simplify(h, cutoff=1.0e-12):
```
Comment (ixfoduap) [inline pennylane/hf/hamiltonian.py:356]: Why not make this visible to users? (So `simplify()`, not `_simplify()`)
Seems like something that could be useful more generally in pennylane.

@josh146 Mentioned today that such general functionality merits thought on where to best implement it. Does it make sense for simplify to exist outside of the `hf` module? Are we simply waiting on the operator refactor to do so?

File: pennylane/hf/hamiltonian.py [hunk 5]
```diff
@@ -352,6 +353,54 @@ def _generate_qubit_operator(op):
...
+    Args:
+        h (Hamiltonian): PennyLane Hamiltonian
+        cutoff (float): cutoff value for discarding the negligible terms
+
+    Returns:
+        Hamiltonian: Simplified PennyLane Hamiltonian
+
+    **Example**
+
+    >>> c = np.array([0.5, 0.5])
```
Comment (ixfoduap) [inline pennylane/hf/hamiltonian.py:371]: Great example! Simple and clear; love it

File: pennylane/hf/hamiltonian.py [hunk 6]
```diff
@@ -352,6 +353,54 @@ def _generate_qubit_operator(op):
...
+    >>> h = qml.Hamiltonian(c, [qml.PauliX(0) @ qml.PauliY(1), qml.PauliX(0) @ qml.PauliY(1)])
+    >>> print(_simplify(h))
+    (1.0) [X0 Y1]
+    """
+    wiremap = dict(zip(h.wires, range(len(h.wires) + 1)))
+
+    c = []
+    o = []
+    for i, op in enumerate(h.terms[1]):
```
Comment (ixfoduap) [inline pennylane/hf/hamiltonian.py:380]: So this is taking care of repeated operators only appearing once, right? And it's done by adding the corresponding coefficients?

File: pennylane/hf/hamiltonian.py [hunk 7]
```diff
@@ -352,6 +353,54 @@ def _generate_qubit_operator(op):
...
+        op = qml.grouping.pauli_word_to_string(op, wire_map=wiremap)
+        if op not in o:
+            c.append(h.terms[0][i])
+            o.append(op)
+        else:
+            c[o.index(op)] += h.terms[0][i]
+
+    coeffs = []
+    ops = []
+    nonzero_ind = np.argwhere(abs(np.array(c)) > cutoff).flatten()
```
Comment (ixfoduap) [inline pennylane/hf/hamiltonian.py:391]: And this means you only keep operators whose coefficient is above a threshold, correct?

File: pennylane/hf/hamiltonian.py (hunk 3 above)
Comment (ixfoduap) [reply pennylane/hf/hamiltonian.py:403]: This PR is literally adding a function that simplifies any input Hamiltonian. As far as I can tell, it should be possible to reproduce in the Hamiltonian class, or as you suggest, having `qml.simplify(H)`. I really like this latter suggestion actually

**Review by ixfoduap — CHANGES_REQUESTED** [review]
The code looks great, no real suggestions.

But I wonder if following Maria's comment we should turn this into a more general `qml.simplify()` function, which I think is a great idea

File: pennylane/hf/hamiltonian.py (hunk 6 above)
Comment (soranjh) [reply pennylane/hf/hamiltonian.py:380]: Yes.

File: pennylane/hf/hamiltonian.py (hunk 7 above)
Comment (soranjh) [reply pennylane/hf/hamiltonian.py:391]: Yes.

File: pennylane/hf/hamiltonian.py (hunk 3 above)
Comment (soranjh) [reply pennylane/hf/hamiltonian.py:403]: Hi @mariaschuld, @ixfoduap and @josh146. Here are my thoughts on the simplify function.

This is a general functions that does not necessary belong to `qml.hf`. However, the way it is written currently is based on one important point: convert Pauli words to strings and manipulate the strings in with python. For example, instead of looping over the words and using `compare`, one can simply use 
`if o in ['XYZ', ...]`
However, I am not sure we want to manipulate Hamiltonians/operators by turning them to strings. Another philosophy could be generating the binary represantation of words and Hamiltonians and manipulate those objects numerically to, for example, compare or multiply them. I think we need to decide first what approach we want to follow for operator manipulation/arithmetic in general. This is the only reason that I hided `_simplify` inside `hf.hamiltonian`.

So, I want your thoughts on these options:
1. Convert `hf.hamiltonian._simplify` to `qml.simplify(H)`
2. Wait until we have an ADR on how we really want to _efficiently_ work with operators/Hamiltonian in PL and then create `qml.simplify(H)` accordingly.

File: pennylane/hf/hamiltonian.py (hunk 4 above)
Comment (soranjh) [reply pennylane/hf/hamiltonian.py:356]: Please see my comment below.

File: pennylane/hf/hamiltonian.py (hunk 3 above)
Comment (josh146) [reply pennylane/hf/hamiltonian.py:403]: @soranjh I think these are very valid points, still a lot needs to be figured out in terms of how to best implement comparison/simplification long-term.

Having said that, it's not quite clear _when_ this research will be done/implemented. Since your version already is an improvement over what is already in PL, I am very much in favour of making it more general now, and continuing to improve it iteratively down the line.

In other words: since this is already a net benefit over what PL users have access to, I don't think there is much disadvantage to making this generally available in the short term?

Comment (mariaschuld) [reply pennylane/hf/hamiltonian.py:403]: @soranjh I would suggest to go for 2 - this is the least work for you now, and it incurs only 100% non-user-facing changes in the future! You are right, Hamiltonian simplification needs an ADR. Also remember that all operators will optionally define terms, and the simplify function should be rather general!

Comment (josh146) [reply pennylane/hf/hamiltonian.py:403]: @mariaschuld makes the following proposal, which I think makes a lot of sense:

1. Continuing working on this PR

2. Once this PR is merged, improve the existing `Hamiltonian.simplify()` method with this code

3. Finally, we can use what we learnt to (a) inform the ADR, and (b) help us decide any final UI/design decisions.

File: pennylane/hf/hamiltonian.py [hunk 8]
```diff
@@ -352,6 +353,54 @@ def _generate_qubit_operator(op):
...
+    for i in nonzero_ind:
+        coeffs.append(c[i])
+        ops.append(qml.grouping.string_to_pauli_word(o[i], wire_map=wiremap))
+
+    try:
+        coeffs = qml.math.stack(coeffs)
+    except ValueError:
```
Comment (obliviateandsurrender) [inline pennylane/hf/hamiltonian.py:398]: When does this `exception` is raised?

**Review by obliviateandsurrender — COMMENTED** [review]
Just leaving a couple of suggestions and clarifications. Everything else looks good to me!

File: doc/releases/changelog-dev.md [hunk 9]
```diff
@@ -4,6 +4,13 @@
 
 <h3>New features since last release</h3>
 
+* Added a modified version of the `_simplify` function to the `hf` module.
+  [(#2103)](https://github.com/PennyLaneAI/pennylane/pull/2103)
+
+  This functions combines redundant terms in a Hamiltonian and eliminates terms with a zero
+  coefficient. The new function makes construction of molecular Hamiltonians more efficient.
```
[the remark this answers is not in the document]
Comment (soranjh) [reply doc/releases/changelog-dev.md:11]: When addition of two terms gives a coefficient of less than `cutoff=1.0e-12`, the terms is just removed from the Hamiltonian. Mentioned the cutoff for clarity.

File: pennylane/hf/hamiltonian.py [hunk 10]
```diff
@@ -247,7 +247,8 @@ def hamiltonian(*args):
         for n, t in enumerate(h_ferm[1]):
 
             if len(t) == 0:
-                coeffs = np.array([h_ferm[0][n]])
+                coeffs = anp.array([0.0])
+                coeffs = coeffs + np.array([h_ferm[0][n]])
                 ops = ops + [qml.Identity(0)]
 
             elif len(t) == 2:
```
[the remark this answers is not in the document]
Comment (soranjh) [reply pennylane/hf/hamiltonian.py:254]: Thanks for the catch. Probably happened when I had different codes for each `if` and then modified them to the same one without noticing it. Does not seem to affect timing. I will correct it with few other similar things and docstring corrections in a new PR.

File: doc/releases/changelog-dev.md [hunk 11]
```diff
@@ -4,6 +4,14 @@
 
 <h3>New features since last release</h3>
 
+* Added a modified version of the `_simplify` function to the `hf` module.
```
Comment (ixfoduap) [inline doc/releases/changelog-dev.md:7]:
```suggestion
* Added a modified version of the `simplify` function to the `hf` module.
```

**Review by ixfoduap — APPROVED** [review]

**Review by obliviateandsurrender — APPROVED** [review]
